src/audio/coreaudio/SDL_coreaudio.m

```diff
@@ -1145,6 +1145,86 @@ output device (in which case we'll try again). */
     return (this->hidden->thread != NULL) ? 0 : -1;
 }
 
+static int
+COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
+{
+    UInt32 alive = 0;
+    pid_t pid = 0;
+
+    AudioObjectPropertyAddress addr = {
+        iscapture ? kAudioHardwarePropertyDefaultInputDevice
+                  : kAudioHardwarePropertyDefaultOutputDevice,
+        iscapture ? kAudioDevicePropertyScopeInput
+                  : kAudioDevicePropertyScopeOutput,
+        kAudioObjectPropertyElementMaster
+    };
+        
+    AudioObjectPropertyAddress nameaddr = {
+        kAudioObjectPropertyName,
+        iscapture ? kAudioDevicePropertyScopeInput
+                  : kAudioDevicePropertyScopeOutput,
+        kAudioObjectPropertyElementMaster
+    };
+    
+    AudioDeviceID devid;
+    OSStatus result = noErr;
+    UInt32 size = 0;
+    
+    // Get the Device ID
+    size = sizeof (AudioDeviceID);
+    result = AudioObjectGetPropertyData(kAudioObjectSystemObject, &addr,
+                                        0, NULL, &size, &devid);
+    
+    if (result != noErr) {
+        return SDL_SetError("%s: Default Device ID not found", "coreaudio");
+    }
+    
+    
+    // Use the Device ID to get the name
+    CFStringRef cfstr = NULL;
+    size = sizeof (CFStringRef);
+    result = AudioObjectGetPropertyData(devid, &nameaddr, 0, NULL, &size, &cfstr);
+    
+    if (result != noErr) {
+        return SDL_SetError("%s: Default Device Name not found", "coreaudio");
+    }
+    
+    CFIndex len = CFStringGetMaximumSizeForEncoding(CFStringGetLength(cfstr),
+                                                    kCFStringEncodingUTF8);
+    
+    char *devname = (char *) SDL_malloc(len + 1);
+    int usable = ((devname != NULL) &&
+              (CFStringGetCString(cfstr, devname, len + 1, kCFStringEncodingUTF8)));
+    CFRelease(cfstr);
+    
+    if (usable) {
+        usable = 0;
+        len = strlen(devname);
+        /* Some devices have whitespace at the end...trim it. */
+        while ((len > 0) && (devname[len - 1] == ' ')) {
+            len--;
+            usable = len;
+        }
+    }
+    
+    if (usable) {
+        devname[len] = '\0';
+    }
+    
+    int numdevices = SDL_GetNumAudioDevices(iscapture);
```

Comment:
For this block, are we able to use AudioObjectGetPropertyData with the devid above, but with kAudioDevicePropertyNominalSampleRate? I'm thinking we can reuse this, maybe...

https://github.com/libsdl-org/SDL/blob/0258ce6cc75fe7e33319544d2add3f2bd4c502f6/src/audio/coreaudio/SDL_coreaudio.m#L155-L175

Reply:
Yeah. I had started something like that but don't know how to get all the information from coreaudio. According to the documentation in SDL_audio.h, it looks like we need sample frequency, number of channels, and sample format right? I managed to get sample frequency and number of channels, however, I couldn't get sample format. I could revert to what I had and try and figure out how to get sample format?

Also I think I might be misunderstanding part of the code, so for clarification: where does the sample format get stored in SDL_AudioSpec?

Reply:
Format is a bitmask, so it'll be a premade format like AUDIO_F32.

Luckily the format field is something we're allowed to skip; any field we can't identify can simply be filled with 0, indicating we couldn't figure out that part of the spec (which I think is what we do for GetAudioDeviceSpec anyway).

Reply:
Perfect, I'll roll that change back to what I had then.